### ISSUE TITLE
Type definitions for 3rd party classes

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#36]
When importing a class with composer, it would be nice to have intellisense suggestions for those classes.

We need a way that users can define an "interface" for a class with its properties and methods -- and their access modifier.

As of #35, PDO should be the first custom type definition.

---

**Blocked by #1 (class instance variables have no suggestions)**

**Comment by TheColorRed** [comment]
This should be fairly easy to implement, Netbeans has already written php files all that is needed to do is copy those file to a local folder and parse those files alongside the initial directory parsing.

If you have Netbeans installed with the php extension they can be found here:

`%NETBEANS_ROOT%\php\phpstubs\phpruntime`

**Comment by TheColorRed** [comment]
I have added 3rd party libraries.

PDO example:
![3rd party libraries](https://cloud.githubusercontent.com/assets/1571806/15153563/427504e2-169f-11e6-85f2-3ba687a3fe89.PNG)

ImageMagick example:
![3rd party libraries 3](https://cloud.githubusercontent.com/assets/1571806/15153721/f6f40b16-169f-11e6-89ae-1b664218e4bb.PNG)


Other types of classes:
![3rd party libraries 2](https://cloud.githubusercontent.com/assets/1571806/15153562/42741596-169f-11e6-84ca-fed5070d9c6c.PNG)

Here is a full list of all the added libraries:

![libraries](https://cloud.githubusercontent.com/assets/1571806/15153678/bafaa0a2-169f-11e6-8220-6e3b0fc47d43.PNG)

**Comment by nevadascout** [comment]
Cool, does this include methods as well or just the class names?

**Comment by TheColorRed** [comment]
Classes, properties, functions and their arguments, phpDoc, etc.

**Comment by nevadascout** [comment]
Awesome, would love to see the code for this!

**Comment by TheColorRed** [comment]
The Server Side:
https://github.com/TheColorRed/crane/blob/3rd-party-libraries/server/src/server.ts#L427

The PHP Files:
https://github.com/TheColorRed/crane/tree/3rd-party-libraries/client/phpstubs

**Comment by TheColorRed** [comment]
If you want to make another project **HvyIndustries/PHPStubs** or something like that, we can then place the files there and download the zip and unzip it from Github when a user wants to install the libraries.

I think they should then be installed to `%APPDATA%/Crane/phpstubs` (and the proper location for osx/linux).

**Comment by nevadascout** [comment]
I've created a repository: https://github.com/HvyIndustries/crane-php-stubs.

Go ahead and submit a pull request for any stubs you've got.

**Comment by TheColorRed** [comment]
Okay I have added the libraries and submitted the pull request:
https://github.com/HvyIndustries/crane-php-stubs/pull/1

**Comment by TheColorRed** [comment]
What do you think is a better idea:

1. Download and unzip the github zip file and save to a folder and read it in
1. Get the contents for each file via a url: https://api.github.com/repos/HvyIndustries/crane-php-stubs/contents/

With option 2 we could create a setting such as `resources` that could have a list of user defined resources for that project.

```json
"crane.resources": {
	"https://api.github.com/repos/HvyIndustries/crane-php-stubs/contents/": "json",
	"/home/user/www-data/php-files/": "dir"
}
```

**Comment by nevadascout** [comment]
I prefer option 1 as it would let you work without needing an internet connection (and is probably faster on launch)

If we make the base folder something like `{home}/Crane/phpstubs/`, then we can have a `/core` subfolder for the main stubs files, and a `/user` subfolder for users to add their own stubs?

**Comment by TheColorRed** [comment]
I have created a new branch (off of optimization-2), stubs can now be downloaded from the repo (crane-php-stubs) and extracted to `$APP_PATH/Crane/phpstubs/crane-php-stubs-master/`. If a user wants to create their own stubs, they can create one or more folders in the `phpstubs` folder and place their stubs in it.

Stub searching is as follows:

```js
glob('/path/to/phpstubs/*/*.php', ()=> { /* Process the stubs */ });
```

**Comment by nevadascout** [comment]
Perfect, nice one!

**Comment by nevadascout** [comment]
While we've added support for PHP's built-in classes and functions with PHP Stubs, this issue was originally created for dependencies installed via composer - so I'll leave it open until we fix those issues.

**Comment by belaustegui** [comment]
Having suggestions for dependencies installed with composer would be wonderful!
